### PR TITLE
Fix Issue #2432 (Attempt 2): move poly.resolve before typevar substitution

### PR DESCRIPTION
**Pull request by abslamp**
Test case and another possible fix on #2432

Move poly.resolve() before typevar substitution in asMemberOf(). 3 versions of asMemberOf are spitted to two parts: before and after substitution. The default asMemberOf() is kept as a wrapper of the two methods to avoid alternation of other code.

Note: this is a quick prototype to prove logic correctness (by trigger the tests), still need to remove some code duplications.

### CODE REVIEW

File: framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java [hunk 1]
```diff
@@ -353,13 +372,54 @@ public static AnnotatedTypeMirror asMemberOf(
...
+                type = asMemberOfImpl(types, atypeFactory, t, elem, type);
+
                 if (!ElementUtils.isStatic(elem)) {
                     atypeFactory.postAsMemberOf(type, t, elem);
                 }
                 return type;
         }
     }
 
+    public static AnnotatedTypeMirror asMemberOfPreSubstitution(
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java]: Can you add javadoc to the modified methods to make it clear what they are doing.

Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java]: `PreSubstitution` sounds wrong - this is doing a substitution. Once there is javadoc of what the method does we can find a better name.

File: framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java [hunk 2]
```diff
@@ -390,8 +450,35 @@ private static AnnotatedTypeMirror asMemberOfImpl(
...
             case UNION:
+            case DECLARED:
+                // return substituteTypeVariables(types, atypeFactory, of, member, memberType);
+                return memberType;
+            default:
+                throw new BugInCF("asMemberOf called on unexpected type.\nt: " + of);
+        }
+    }
+
+    private static AnnotatedTypeMirror asMemberOfImpl(
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java]: Again, javadoc to make clear what the method does.
Why is this extra method needed? Instead, couldn't you split up the original method into two separate methods?

File: framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java (hunk 1 above)
Comment (abslamp) [reply framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java]: Maybe the asMemberOfImpl is making the name confusing. Actually it does not do a substitution. So I will make asMemberOfImpl more clear.

**Comment by abslamp** [comment]
The last commit is reversed to make the PR clearer.

File: checker/tests/nullness/Issue2432.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
+// Test case for issue 2432:
```
Comment (wmdietl) [inline checker/tests/nullness/Issue2432.java]: Above this in the PR I see the addition of two files: `checker/annotation-tools` and `checker-stubparser`.
However, I can't edit nor delete those files.
Can you remove those changes?

File: checker/tests/nullness/Issue2432.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+// https://github.com/typetools/checker-framework/issues/2432
+
+import java.util.List;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.checkerframework.checker.nullness.qual.PolyNull;
+
+class Issue2432 {
```
Comment (wmdietl) [inline checker/tests/nullness/Issue2432.java]: Can you clean up this test case a bit?
Add a bit of structure and clean naming. Explain a bit (very briefly) what is tested in each section.

File: framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 5]
```diff
@@ -1609,6 +1613,7 @@ protected AnnotatedDeclaredType getImplicitReceiverType(ExpressionTree tree) {
                 if (path == null) {
                     // The path is null if the field is in a compilation unit we haven't
                     // processed yet. TODO: is there a better way?
+                    // This only arises in the Nullness Checker when substituting rawness.
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: Why did you add this comment? What do you mean with `rawness`?

File: framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 6]
```diff
@@ -1211,7 +1211,11 @@ private AnnotatedTypeMirror fromExpression(ExpressionTree tree) {
 
         AnnotatedTypeMirror result = TypeFromTree.fromExpression(this, tree);
 
-        if (shouldCache) {
+        if (shouldCache && tree.getKind() != Tree.Kind.NEW_CLASS) {
+            // Don't cache types of NEW operator
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: How about:

```
// Don't cache the type of object creations, because incorrect
// annotations would be cached during dataflow analysis.
// See Issue #602.
```

File: framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java [hunk 7]
```diff
@@ -223,6 +223,10 @@ public GenericAnnotatedTypeFactory(BaseTypeChecker checker, boolean useFlow) {
             flowResultAnalysisCaches = null;
         }
 
+        // Add common aliases.
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java]: What is the point of this change?

File: framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java [hunk 8]
```diff
@@ -1579,10 +1583,19 @@ public ParameterizedExecutableType methodFromUse(MethodInvocationTree tree) {
...
             dependentTypesHelper.viewpointAdaptMethod(tree, method);
         }
-        poly.resolve(tree, method);
+
         return mType;
     }
 
+    @Override
+    public void methodFromUsePreSubstitution(ExpressionTree tree, AnnotatedTypeMirror mirror) {
+        // no call to super needed since super is no-op
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java]: Well, there is a call to super on the next line... why assume that the behavior in the super method will never change?

File: framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java [hunk 9]
```diff
@@ -1579,10 +1583,19 @@ public ParameterizedExecutableType methodFromUse(MethodInvocationTree tree) {
...
+        super.methodFromUsePreSubstitution(tree, mirror);
+        if (tree instanceof MethodInvocationTree && mirror instanceof AnnotatedExecutableType) {
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java]: What other trees/types could be passed in here?
The method name makes it sound like this would only ever make sense for these more specific types.

File: framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java [hunk 10]
```diff
@@ -364,8 +392,9 @@ private static AnnotatedTypeMirror asMemberOfImpl(
             final Types types,
             final AnnotatedTypeFactory atypeFactory,
             final AnnotatedTypeMirror of,
-            final Element member) {
-        AnnotatedTypeMirror memberType = atypeFactory.getAnnotatedType(member);
+            final Element member,
+            AnnotatedTypeMirror memberType) {
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java]: Can you document the method and additional parameter?

File: framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java (hunk 5 above)
Comment (abslamp) [reply framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: That's odd... It's not added by me, I will check whether there is any merge issues on my branch.

File: framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java (hunk 7 above)
Comment (abslamp) [reply framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java]: Still not added by me

File: framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java (hunk 8 above)
Comment (abslamp) [reply framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java]: I will remove this line

File: checker/tests/nullness/Issue2432.java (hunk 3 above)
Comment (abslamp) [reply checker/tests/nullness/Issue2432.java]: I cannot remove either, every time I tried to delete or edit github would return a 404 page.

**Comment by abslamp** [comment]
Pull request closed due to uncorrectable or historical issues.
